spec/salestation/web/extractors/body_param_extractor_spec.rb

```diff
@@ -176,4 +176,132 @@ def initialize(env_data)
       end
     end
   end
+
+  context 'with filters as hashes' do
+    subject(:extract_body_params) { described_class[:x, {a: [:b, :c], d: [:e, :f]}].call(request) }
+
+    let(:params) { {'x' => 0, 'a' => { 'b' => 1, 'c' => 2}, 'd' => {'e' => 3, 'f' => 4, 'g' => 5}} }
+    let(:expected_result) { {x: 0, a: { b: 1, c: 2}, d: {e: 3, f: 4, g: 5} } }
```

Comment:
On line 181 it was explicitly marked that `d: [:e, :f]` is whitelisted, but `g` got whitelisted as well?

Reply:
Thank you for pointing it out. This test was made such a way in purpose because I had the opinion last week that fully supporting the nested whitelisting would break some of current behaviour when the algorithm is fully recursive. However, this week, having meditated on this example once more, I have belief that this can still be handled and I will try to do that.

I am a bit suspicious, if filtering the HTTP input based on this whitelisting is actually the best way to go. To me it would make more sense, if business logic inside application would "use" only the part of input that is necessary and nothing would be filtered out before reaching the actual business logic. Then the decision what to use and what not, would be in a single place. But this is not relevant at the moment. If it has been done like this so far, then  yes, I will try to continue supporting it fully.

Reply:
> I am a bit suspicious, if filtering the HTTP input based on this whitelisting is actually the best way to go. To me it would make more sense, if business logic inside application would "use" only the part of input that is necessary and nothing would be filtered out before reaching the actual business logic.

There are many inputs. Query params, Body params, Headers, Path. Imagine if query parameters has `foo` and body parameter also has parameter `foo` with a different value. Who wins? We also don't want to send "web" specific information to the application. Sometimes the request is done by something else that an HTTP request (although it's rare). We've so far tried to keep the application code independent of the UI / web layer. So it requires inputs, and the web layer is responsible for providing those inputs. Now, how explicit these have to be, is a different question, but so far being explicit has worked quite well.

My last comment in the PR actually agrees what you're doing, with couple of small changes.